packages/flutter/lib/src/foundation/diagnostics.dart

```diff
@@ -3677,3 +3682,38 @@ class _DefaultDiagnosticsSerializationDelegate implements DiagnosticsSerializati
     );
   }
 }
+
+/// A value that can be weakly held.
+@visibleForTesting
+class WeakValue<T> {
+
+  /// Creates a weak value.
+  ///
+  /// If [keepAlive] is true, then the value will be strongly held.
+  WeakValue(T? value, {bool keepAlive = false}) {
+    if (value == null) {
+      return;
+    }
+
+    // If weak reference does not support T, then use strong reference.
+    // See https://api.dart.dev/stable/3.0.2/dart-core/WeakReference-class.html
+    if (keepAlive || value is String || value is num || value is bool || value is Record) {
```

Comment:
special casing null, String, num, and bool make sense.
For other cases like records it is probably ok to throw and just note that only types that are valid for Expando are supported. That would be less surprising than leaking for those cases.